diff --git a/Android.mk b/Android.mk
index 78b9430..8115908 100755
--- a/Android.mk
+++ b/Android.mk
@@ -360,6 +360,14 @@ ifeq ($(TW_INCLUDE_CRYPTO), true)
         endif
         LOCAL_SHARED_LIBRARIES += libcryptfs_hw
     endif
+    ifeq ($(TW_INCLUDE_OMAPI),true)
+        LOCAL_CFLAGS += -DTW_INCLUDE_OMAPI
+        LOCAL_SHARED_LIBRARIES += android.hardware.secure_element-V1-ndk android.se.omapi-V1-ndk
+        TWRP_REQUIRED_MODULES += \
+            se_omapi \
+            se_omapi.rc \
+            se_omapi.xml
+    endif
     TW_INCLUDE_LIBRESETPROP := true
 endif
 WITH_CRYPTO_UTILS := \
diff --git a/etc/Android.mk b/etc/Android.mk
index 8bd900d..9d9f243 100755
--- a/etc/Android.mk
+++ b/etc/Android.mk
@@ -205,3 +205,21 @@ ifeq ($(TW_USE_TOOLBOX), true)
     LOCAL_SRC_FILES := $(LOCAL_MODULE)
     include $(BUILD_PREBUILT)
 endif
+
+ifeq ($(TW_INCLUDE_OMAPI),true)
+    include $(CLEAR_VARS)
+    LOCAL_MODULE := se_omapi.rc
+    LOCAL_MODULE_TAGS := optional
+    LOCAL_MODULE_CLASS := ETC
+    LOCAL_MODULE_PATH := $(TARGET_RECOVERY_ROOT_OUT)/system/etc/init
+    LOCAL_SRC_FILES := init/$(LOCAL_MODULE)
+    include $(BUILD_PREBUILT)
+
+    include $(CLEAR_VARS)
+    LOCAL_MODULE := se_omapi.xml
+    LOCAL_MODULE_TAGS := optional
+    LOCAL_MODULE_CLASS := ETC
+    LOCAL_MODULE_PATH := $(TARGET_RECOVERY_ROOT_OUT)/vendor/etc/vintf/manifest
+    LOCAL_SRC_FILES := init/$(LOCAL_MODULE)
+    include $(BUILD_PREBUILT)
+endif
diff --git a/prebuilt/Android.bp b/prebuilt/Android.bp
index fdbb3d2..985c5c7 100644
--- a/prebuilt/Android.bp
+++ b/prebuilt/Android.bp
@@ -1,5 +1,6 @@
 soong_namespace {
     imports: [
 		"external/lptools",
+		"external/se_omapi",
 	],
 }
diff --git a/prebuilt/Android.mk b/prebuilt/Android.mk
index f3b7160..b23ba17 100755
--- a/prebuilt/Android.mk
+++ b/prebuilt/Android.mk
@@ -241,6 +241,11 @@ ifeq ($(TW_INCLUDE_CRYPTO), true)
         RECOVERY_LIBRARY_SOURCE_FILES += $(TARGET_OUT_SHARED_LIBRARIES)/libhardware_legacy.so
         RECOVERY_LIBRARY_SOURCE_FILES += $(TARGET_OUT_SHARED_LIBRARIES)/android.system.suspend@1.0.so
         RECOVERY_BINARY_SOURCE_FILES += $(TARGET_OUT_EXECUTABLES)/fscryptpolicyget
+        ifeq ($(TW_INCLUDE_OMAPI),true)
+            RECOVERY_BINARY_SOURCE_FILES += $(TARGET_OUT_EXECUTABLES)/se_omapi
+            RECOVERY_LIBRARY_SOURCE_FILES += $(TARGET_OUT_SHARED_LIBRARIES)/android.se.omapi-V1-ndk.so
+            RECOVERY_LIBRARY_SOURCE_FILES += $(TARGET_OUT_SHARED_LIBRARIES)/android.hardware.secure_element-V1-ndk.so
+        endif
         ifneq ($(TW_EXCLUDE_LIBXML2), true)
             RECOVERY_LIBRARY_SOURCE_FILES += $(TARGET_OUT_SHARED_LIBRARIES)/libxml2.so
         endif
