diff --git a/minuitwrp/events.cpp b/minuitwrp/events.cpp
--- a/minuitwrp/events.cpp
+++ b/minuitwrp/events.cpp
@@ -159,7 +159,7 @@ int vibrate(int timeout_ms)
         vib->on((uint32_t)timeout_ms);
     }
 #elif defined(USE_QTI_AIDL_HAPTICS)
-    std::shared_ptr<IVibrator> vib = IVibrator::fromBinder(ndk::SpAIBinder(AServiceManager_getService(kVibratorInstance.c_str())));
+    std::shared_ptr<IVibrator> vib = IVibrator::fromBinder(ndk::SpAIBinder(AServiceManager_checkService(kVibratorInstance.c_str())));
     if (vib != nullptr) {
 #ifdef USE_QTI_AIDL_HAPTICS_FIX_OFF
         std::thread([vib, timeout_ms] {
